### PR TITLE
SAK-46034 Lessons Add templates for page layouts

### PR DESCRIPTION
**Pull request by jkozar2**
This PR contains a system that gives users the option to add multi-item page-layout templates to Lessons. There are currently three templates, each with a description and preview image; the system is easily expandable to include more templates in the future. The templates are modular; more than one can be put on a page, and a newly-added one will tack onto the bottom of a page without affecting what's already present.

### CODE REVIEW

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 1]
```diff
@@ -1143,6 +1154,69 @@ public void setLayoutSelect(String layoutSelect) {
 		this.layoutSelect = layoutSelect;
 	}
 
+	public void setPageLayoutSelect(String pageLayoutSelect) {
```
Comment (ern) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:1157]: lombok annotations will make this boiler plate go away!

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 2]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			itemNow = simplePageToolDao.makeItem(getCurrentPageId(),sequence,SimplePageItem.BREAK,null,null);
+			sequence++;
+			itemNow.setFormat("column");
+			itemNow.setAttribute("forceBtn","false");
+			itemNow.setAttribute("checklistItems",null);
+			itemNow.setAttribute("colcolor","trans");
+			update(itemNow);	//finish writing third item, Break for the checklist
+			itemNow = simplePageToolDao.makeItem(getCurrentPageId(),sequence,SimplePageItem.CHECKLIST,null,"Checklist");
+			sequence++;
+			itemNow.setDescription("Track your progress");
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4761]: Can you provide i18n translations for this?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 3]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			itemNow.setAttribute("indentLevel","0");
+			itemNow.setAttribute("namehidden","false");
+			simplePageToolDao.addChecklistItem(itemNow,Long.valueOf(0),"Add item here",Long.valueOf(-1));
+			simplePageToolDao.addChecklistItem(itemNow,Long.valueOf(1),"Add item here",Long.valueOf(-1));
+			simplePageToolDao.addChecklistItem(itemNow,Long.valueOf(2),"Add item here",Long.valueOf(-1));
```
Comment (bjones86) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4766]: And here, and any others that may have been missed

**Review by bjones86 — CHANGES_REQUESTED** [review]

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 4]
```diff
@@ -393,6 +402,8 @@
 			"Teal",
 			"Purple"
 	};
+	public static final String pageLayoutValues[] = {"none", "addSubpageList", "interiorResources", "interiorTask"};
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:405]: Could we use an ENUM here?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 5]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			if(itemNow.getSequence() > currentHighest){
+				currentHighest = itemNow.getSequence();
+			}
+		}
+		return currentHighest + 1;
+	}
+
+	public String addPageLayout(){
+		int sequence = getNextSequenceOnPage(getCurrentPageId());	//find the starting sequence number based on whatever might already be on the page. Must be incremented after every use.
+       	if(StringUtils.equals(this.pageLayoutSelect, "addSubpageList")){
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4702]: Could we use a switch instead using the ENUM?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 6]
```diff
@@ -393,6 +402,8 @@
 			"Teal",
 			"Purple"
 	};
+	public static final String pageLayoutValues[] = {"none", "addSubpageList", "interiorResources", "interiorTask"};
+	public static final String subpageCountValues[] = {"1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20"};
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:406]: Not a big fan of this, could we maybe use something like IntStream.range(1, 21).toArray()

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 7]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+	}
+
+	public String addPageLayout(){
+		int sequence = getNextSequenceOnPage(getCurrentPageId());	//find the starting sequence number based on whatever might already be on the page. Must be incremented after every use.
+       	if(StringUtils.equals(this.pageLayoutSelect, "addSubpageList")){
+			SimplePage pageNow = null;
+			SimplePageItem itemNow = null;
+			if(StringUtils.isBlank(pageSubpageTitle)){
+				setErrMessage(messageLocator.getMessage("simplepage.layout.page.emptyName"));
+				return "failure";
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4707]: Could we use constants for "failure"?

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 8]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			}
+			itemNow = simplePageToolDao.makeItem(getCurrentPageId(),sequence,SimplePageItem.BREAK,null,null);
+			sequence++;
+			itemNow.setFormat("section");
+			itemNow.setAttribute("forceBtn","false");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4712]: Boolean.FALSE.toString()

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 9]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+				itemNow.setSequence(sequence);
+				sequence++;
+				update(itemNow);	//save revised item data
+			}
+		} else if(StringUtils.equals(this.pageLayoutSelect, "interiorResources") || StringUtils.equals(this.pageLayoutSelect, "interiorTask")){	//first handle the parts that these layouts have in common.
+			SimplePageItem itemNow = simplePageToolDao.makeItem(getCurrentPageId(), sequence, SimplePageItem.BREAK, null, null);
+			sequence++;
+			itemNow.setFormat("section");
+			itemNow.setAttribute("colwidth", "2");
+			itemNow.setAttribute("forceBtn","false");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4735]: Boolean.FALSE.toString()

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java (hunk 8 above)
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4712]: Let's define a constant FORCE_BTN and reuse it

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 10]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+				itemNow.setPageId(getCurrentPageId());
+				itemNow.setAttribute("btnColor",pageButtonColorScheme);
+				itemNow.setSequence(sequence);
+				sequence++;
+				update(itemNow);	//save revised item data
+			}
+		} else if(StringUtils.equals(this.pageLayoutSelect, "interiorResources") || StringUtils.equals(this.pageLayoutSelect, "interiorTask")){	//first handle the parts that these layouts have in common.
+			SimplePageItem itemNow = simplePageToolDao.makeItem(getCurrentPageId(), sequence, SimplePageItem.BREAK, null, null);
+			sequence++;
+			itemNow.setFormat("section");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4733]: SECTION constant

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 11]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			itemNow.setAttribute("colwidth", "2");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4734]: Use constants

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 12]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+			itemNow.setAttribute("forceBtn","false");
+			itemNow.setAttribute("checklistItems",null);
+			update(itemNow);	//finish writing top Section/break item
+			itemNow = simplePageToolDao.makeItem(getCurrentPageId(), sequence,SimplePageItem.TEXT,null,null);
+			sequence++;
+			String overviewHtmlAppender = "";
+			if(StringUtils.equals(this.pageLayoutSelect, "interiorResources")){	//special Overview text for interiorResources only
+				overviewHtmlAppender = overviewHtmlAppender + messageLocator.getMessage("simplepage.layout.page.overview","Tasks");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4742]: Is this hardcoded in English? I'm fine raising I18N or L10N bugs after merging this.

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java [hunk 13]
```diff
@@ -4609,14 +4683,131 @@ private String uploadFile(String collectionId) {
...
+				overviewHtmlAppender = overviewHtmlAppender + messageLocator.getMessage("simplepage.layout.page.overview.lo");
+			}
+			overviewHtmlAppender = overviewHtmlAppender + messageLocator.getMessage("simplepage.layout.page.overview.tail");
+			itemNow.setHtml(overviewHtmlAppender);
+			itemNow.setAttribute("checklistItems",null);
+			update(itemNow);	//finish writing second item [overview]
+			itemNow = simplePageToolDao.makeItem(getCurrentPageId(),sequence,SimplePageItem.BREAK,null,null);
+			sequence++;
+			itemNow.setFormat("column");
+			itemNow.setAttribute("forceBtn","false");
```
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4755]: Use constants

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java (hunk 2 above)
Comment (mpellicer) [inline lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:4761]: Is this hardcoded in English? I'm fine raising I18N bugs later.

File: lessonbuilder/tool/src/webapp/js/show-page.js [hunk 14]
```diff
@@ -3978,3 +3978,71 @@ function fixStatus(here,itemId) {
 	return;
     };
 }
+
+function pageLayoutSelectChange(){
+	$(".templateTypeDescription").hide();  //hide all the descriptions in case one is already showing
+	$("#simplepage-layout-example").hide();
+	var value = document.getElementById('page-dropdown-selection').value;
+	if(value === 'addSubpageList'){    //show only the description we've selected.
```
Comment (mpellicer) [inline lessonbuilder/tool/src/webapp/js/show-page.js:3986]: Could we use switch instead?

**Comment by mpellicer** [comment]
I made around 13 comments but honestly, it's about source and not about behavior, I really appreciate this amazing contribution @jkozar2 and I'm not going to block the merge for my comments!

@ern I don't think we should block the PR for our comments, if they can address them better but definitelly a good contribution!

File: lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java (hunk 6 above)
Comment (jkozar2) [reply lessonbuilder/tool/src/java/org/sakaiproject/lessonbuildertool/tool/beans/SimplePageBean.java:406]: I have remade the array where it gets used, in ShowPageProducer.createPageLayoutTab, with a loop if that works better.

**Review by bjones86 — APPROVED** [review]

**Pull request merged**